plugins/Diagnostics/ConfigReader.php

```diff
@@ -179,19 +179,21 @@ public function addConfigValuesFromSystemSettings($configValues, $systemSettings
                 if (isset($configValues[$configSection][$name])) {
                     $configValues[$configSection][$name]['defaultValue'] = $setting->getDefaultValue();
                     $configValues[$configSection][$name]['description']  = trim($description);
-
-                    if ($config->uiControl === PiwikSettings\FieldConfig::UI_CONTROL_PASSWORD) {
-                        $configValues[$configSection][$name]['value'] = $this->getMaskedPassword();
-                    }
                 } else {
-                    $defaultValue = $setting->getValue();
                     $configValues[$configSection][$name] = array(
-                        'value' => null,
+                        'value' => $setting->getValue() != $setting->getDefaultValue() ? $setting->getValue() : null,
                         'description' => trim($description),
-                        'isCustomValue' => false,
-                        'defaultValue' => $defaultValue
+                        'isCustomValue' => $setting->getValue() != $setting->getDefaultValue(),
+                        'defaultValue' => $setting->getDefaultValue()
                     );
                 }
+
+                if (
+                    !empty($configValues[$configSection][$name]['value'])
+                    && $config->uiControl === PiwikSettings\FieldConfig::UI_CONTROL_PASSWORD
```

Comment:
Haven't done any benchmarking and don't know PHP internals good enough to be able to say from top of my head, but would the comparison for uiControl type be better as the first condition compared to 3 levels down array value lookup? Perhaps no difference there, but just crossed my mind that in general it's best to put the conditions that can cause skipping the execution most likely at the beginning so that the rest doesn't need to be evaluated at all.